### ISSUE TITLE
Compatibility with Name Your Price - how to apply currency conversion to min/max prices

### ISSUE THREAD

**Issue by helgatheviking** [Automattic/woocommerce-payments#3507]
### Describe the bug
A $1 price is entered in the default USD currency, but the cart is switched to australian dollars AUD and the product's price in the cart becomes $2... which is not what I'd expect given the conversion rate is 1 USD = 1.4 AUD. 

### To Reproduce

1. Create a Name Your Price simple product
2. Enable WooCommerce Payments multicurrency feature and at least 1 additional currency
3. Add the NYP product to the cart
4. Switch currency

### Actual behavior
Cart price is neither the same as was entered, nor is it converted per currency exchange rates. It seems to do something entirely random.

### Screenshots
![image](https://user-images.githubusercontent.com/507025/145617150-f9d6d681-6719-4a4f-baa4-80816512d8a1.png)

![image](https://user-images.githubusercontent.com/507025/145617272-f541f30e-0e5f-4746-9f9c-1bc901cdfa33.png)

![image](https://user-images.githubusercontent.com/507025/145617364-c45dbd01-f1e2-4d93-954b-c5e58064e27e.png)

### Desktop (please complete the following information):

* OS: Windows 11
* Browser Firefox
* Version 95

I suspect we will need some compatibility code to handle this, but am posting it in this repo in case you can give me some insight into the multicurrency feature and how it changes prices.

**Comment by helgatheviking** [comment]
I apologize in advance that I am probably going to stream of consciousness update this issue as I work my way through it (and learn a little more about the multicurrency feature). :)

**Update 1** - The funky prices are due to rounding settings that I just discovered. So NYP's entered price is actually converted automagically... woo!

**Update 2 -** That means the issue my customer is having more related to minimum price validation. Their store default currency is AUD. They have a product with a $1 minimum. However, what is happening is that entering $1 and then switching to USD (let's assume a 1 AUD = .7 USD for clarity) means the entered price of $1 becomes $0.70 dollars.... which they are thinking is below the minimum. 

Currently NYP is only validating the entered amount.. and not re-validating the converted amount. But even if it did, shouldn't it re-validate against a converted minimum. So a $1 AUD min would be $0.70 USD according to the set rate. 

this plugin doesn't seem to provide for per-currency fixed prices, is that correct? (the admin price entered is always converted?) So what I think I need to do here, is convert the minimum price to the chosen currency. I see there is a method for [converting a price](https://github.com/Automattic/woocommerce-payments/blob/develop/includes/multi-currency/FrontendPrices.php#L79), is there a way to call that from outside the plugin ? 

I also see there's a condition to check IF the price should be converted... if their preference was to _not_ convert the entered amount... is there a filter I can use to enable that?

Would love to hear your thoughts on how we  _should_ we handle entered prices? Should they get converted? And should they get validated against the converted min/max prices? I think that's what we're doing for compat with the WooCommerce Multicurrency plugin.

**Comment by helgatheviking** [comment]
**Update 3 -** We've found that if you are not in the default currency when you add the product to the cart, the amount will be converted from the default currency in the cart:

Default currency is AUD

$1 entered while site is in USD
![image](https://user-images.githubusercontent.com/507025/145633999-6571f478-66bf-48ea-9de6-2ac72f140beb.png)

The cart price should stay as $1 USD. However, $1 is converted from AUD to USD:
![image](https://user-images.githubusercontent.com/507025/145634138-ce6d95ac-fd75-4ed5-85be-46f6bb95eb14.png)

So this opens up 2 needs...
1. need to filter the min/max prices `_min_price` and `_maximum_price` meta keys or via the `wc_nyp_raw_minimum_price` and `wc_nyp_raw_maximum_price` filters.

2. Prevent conversion on products entered in current currency. NYP is setting the price on the product object in the cart directly `$product->set_price( $cart_item['nyp'] );` I am guessing that Payments is coming along on the `woocommerce_product_get_price` filter and converting the price I just set.

**Comment by vbelolapotkov** [comment]
Hey @helgatheviking ! Thanks for submitting this issue. Glad you've discovered rounding settings which were the source of initial confusion. I'll ask folks who worked on this feature to take a closer look to your questions. In the meantime, since the initial issue (in description) seems to be solved by rounding settings, could you please write the summary of what you're trying to achieve and where are you blocked? Asking this because it's a bit hard to follow the "Updates" comments above and evaluate your needs 🙂

**Comment by ismaeldcom** [comment]
Hi @helgatheviking! I'll try to answer your questions and provide more details on how MultiCurrency works 🙂

The first thing to mention is that Name Your Price plugin is not compatible with MultiCurrency. Prices are not currently converted, only formatting is applied. Taking as example USD and JPY, to see the difference clearly, $1 should convert to ¥200 with the default settings, but it shows like ¥1 instead.

That's why the minimum amount is not working as expected. To make it compatible, like you already mentioned, hooking into all `wc_nyp_raw_*_price` filters is needed. But that will work only for the fronted part, making it also compatible in the checkout will require additional work.


> this plugin doesn't seem to provide for per-currency fixed prices, is that correct? (the admin price entered is always converted?) So what I think I need to do here, is convert the minimum price to the chosen currency. I see there is a method for converting a price, is there a way to call that from outside the plugin ?

Yes, for now, all prices in admin will be converted using the rate configured in Multi-Currency settings. As I mentioned before, the best approach is to handle this with compatibility code. You can check some examples looking at [current compatibility code](https://github.com/Automattic/woocommerce-payments/tree/develop/includes/multi-currency/Compatibility).

> Would love to hear your thoughts on how we should we handle entered prices? Should they get converted? And should they get validated against the converted min/max prices? I think that's what we're doing for compat with the WooCommerce Multicurrency plugin.

Hooking in the mentioned filters will handle the product part right. But for the checkout there are some more things to do, like disabling the formatting rules, otherwise a $1 price could end like $0.99 in the cart. A good example will be [the Product AddOns compatibility code](https://github.com/Automattic/woocommerce-payments/blob/develop/includes/multi-currency/Compatibility/WooCommerceProductAddOns.php).

Hope it helps to better understand how it works and solve your case. Let us know if you have any other questions!

**Comment by helgatheviking** [comment]
>The first thing to mention is that Name Your Price plugin is not compatible with MultiCurrency. 

Yup, found that out when I was opening this issue and handling the customer in support. I didn't even know y'all had this feature! :)

I'll try to take a look at the Addons code when I can. With other plugins what I usually do is add the user's currency code to the cart item array, and then skip any conversions as long as current currency match the currency at the time the customer added the item to the cart. Then if the currency changes we convert from the entered amount/currency to the current currency. 

I'm not sure I fully understand this meta key you have: https://github.com/Automattic/woocommerce-payments/blob/develop/includes/multi-currency/Compatibility/WooCommerceProductAddOns.php#L61 can you elaborate on it a bit?

**Comment by ismaeldcom** [comment]
Sure! That meta key is used to know that the cart items have already been converted. You can see how we set it to `1` in `update_product_price` [here](https://github.com/Automattic/woocommerce-payments/blob/7c96f0325a337bb6eb864a55eab91af036fc1870/includes/multi-currency/Compatibility/WooCommerceProductAddOns.php#L210). It's needed cause **Product AddOns** stores prices in meta, and they need to be converted only once.

**Name Your Price** do something similar by storing the price entered in the `nyc` meta. It should be converted to the target currency only once.

The **Product AddOns** may be a bit complex, because it needs some more work to be compatible. Although is a good example, you can start with a simpler one. A good starting point will be something like:

```php
<?php
namespace WCPay\MultiCurrency\Compatibility;

class WooCommerceNameYourPrice extends BaseCompatibility {

	protected function init() {
		if ( class_exists( 'WC_Name_Your_Price' ) ) {
			add_filter( 'wc_nyp_raw_minimum_price', [ $this, 'wc_nyp_prices' ] );
			add_filter( 'wc_nyp_raw_maximum_price', [ $this, 'wc_nyp_prices' ] );
			add_filter( 'wc_nyp_raw_suggested_price', [ $this, 'wc_nyp_prices' ] );
		}
	}

	public function wc_nyp_prices( $price ): float {
		return $this->multi_currency->get_price( $price, 'product' );
	}
}
```

Also, be sure to add the new compatibility class [here](https://github.com/Automattic/woocommerce-payments/blob/7c96f0325a337bb6eb864a55eab91af036fc1870/includes/multi-currency/Compatibility.php#L55).

**Comment by helgatheviking** [comment]
Ah got it! so that meta is simply a flag. is there risk of that being saved? I've been wondering about that myself for a while.... currently NYP uses `set_price()` on the cart's product object instead of filtering prices (as we're only modifying the price in the cart). 

In the case of NYP though, we need to be careful to not always convert from the DB currency. I think we should convert from the currency at the time the customer entered the price and added it to the cart. At least this is what we do in the compat with the Multicurrency plugin and others. 

Assume database/store prices are in USD.
Customer switches to EUR
Enters 100EUR and adds to cart
in cart, switches back to USD
I would expect the cart price to now be 112 USD.

Still wrapping my head around the plugin obviously, but afaik, it looks like it always tries to convert from the base currency to current? is that correct? I don't think that's quite right from my case, unless you can think of a better way to handle customer's entering prices in different currencies?

**Comment by ismaeldcom** [comment]
Yes, **Multi-Currency** will convert the store currency to the user-selected one by default. You can prevent that using the `wcpay_multi_currency_ should_convert_product_price` filter.

But for this case, among different options, maybe the simplest one is to just convert the input amount and store it using the store currency. So it gets properly converted all the time. If I'm not missing anything, the starting point code I mentioned before and hooking into NYC set price should work.

I don't know **Name Your Price** deeply enough, but it could be a way to avoid hooking into many filters and actions.

Your previous example will look like:
- Store in USD
- Customer switches to EUR
- Enters 100 EUR and adds to cart
- Price stored as 112 USD
- In cart will be converted and show 100 EUR
- Switches to USD
- No conversion, it will show 112 USD

**Comment by helgatheviking** [comment]
Thanks @ismaeldcom .... I know Name Your Price and you know Multicurrency, so hopefully we can come up with a solution together. :)

In your above example, I think the part that I am missing is how to convert from X currency to Y currency? 

- Enters 100 EUR and adds to cart
- Price stored as 112 USD

So far, I have only been able to work out that you are converting from the store currency to the current currency (so only ever USD to EUR in this example). 

Multicurrency's [`get_price()`](https://github.com/Automattic/woocommerce-payments/blob/develop/includes/multi-currency/MultiCurrency.php#L717) method doesn't appear to have a parameter for assigning to and from currencies, so how do to convert from EUR to USD?

I _think_ that should be enough for me to implement the steps you have suggested.

**Comment by ismaeldcom** [comment]
Anytime! 🙂 

In this case, we don't want to use `get_price()` because it will apply charming and rounding formatting rules. Instead, you can get the user selected currency with [`get_selected_currency()`](https://github.com/Automattic/woocommerce-payments/blob/8f140393a3f964659b373275b762855413e6016a/includes/multi-currency/MultiCurrency.php#L613) and use the `rate` property to do the conversion.

You can see how we do it inside `get_price()`:
https://github.com/Automattic/woocommerce-payments/blob/8f140393a3f964659b373275b762855413e6016a/includes/multi-currency/MultiCurrency.php#L725

Hope it helps to do an initial implementation!

**Comment by helgatheviking** [comment]
Is there any chance someone could look at my PR #3578 ? I have a mutual customer in support today needing this.

**Comment by ismaeldcom** [comment]
Hi @helgatheviking, I added Sigma team as reviewer, and will take a look at it soon.

**Comment by helgatheviking** [comment]
Do you know of any way to make the compat class a standalone sort of plugin that could perhaps work as a temp patch while the PR is under review? Specifically I am not sure how instantiate the class with the constructor variables.... here : https://github.com/Automattic/woocommerce-payments/pull/3578/files#diff-d23b7ff205ceb290f7dc0543e3845909c3361436eb26c45c672f58f9b9731522R60

**Comment by ismaeldcom** [comment]
You can see in the following code how we instantiate it: https://github.com/Automattic/woocommerce-payments/blob/6fad5475b3ca5560112a427fe302836230a73609/includes/multi-currency/wc-payments-multi-currency.php#L15-L19

One option could be instantiating your plugin after WC Payments Multi Currency, and using `WC_Payments_Multi_Currency()` to access it to initialize your compatibility class.

**Comment by helgatheviking** [comment]
Thanks Ismael. Is that the same process for the `$this->utils` that is also used in the constructor of the compat class?

**Comment by ismaeldcom** [comment]
Not the same, as `$utils` is protected, and we're not exposing it nor have any hook, you will need to initialize a new one from `WooCommerce\Payments\Utils`. But I don't see any issue on doing it for a temporary solution.
